### DISCUSSION TITLE
How to use integrate typescript?

### DISCUSSION THREAD

**Discussion by Altair2169** [suren-atoyan/monaco-react#320] [Q&A]
I took a look at the types file in `src` but I'm still not clear on how to integrate typescript with it. 
for eg. Here's my  `onMount` function. 

```
const handleEditorDidMount = (editor, monaco) => {
    editorRef.current = editor;
    monacoRef.current = monaco;
    console.log(monacoRef.current.languages.getLanguages());
    console.log('editor', editor.getModel().getLanguageIdentifier().language);
  };
```
My question is what should be the types of `editor` and `monaco` in the function arguments?

**Comment by Altair2169** [comment]
I installed monaco editor as a dev dependency and tried to include `node_modules/monaco-editor/monaco.d.ts` in the `include` block in `tsconfig.json` but it still threw the error `cannot find namespace 'monaco'` #321. The only way I got it to work is to  directly import monaco editor like this
`import * as monaco from 'monaco-editor'`

**Comment by ELLIOTTCABLE** [comment]
A little bit cleaner way to type the `editor` and `monaco` arguments is to apply the exported handler-types to your handlers:

```typescript
import MonacoEditor from "@monaco-editor/react"
import type { BeforeMount, OnMount } from "@monaco-editor/react"

// ...

export default function Editor(props: Props) {
   const handleEditorWillMount: BeforeMount = (monaco) => { monaco.foo... }

   const handleEditorDidMount: OnMount = (editor, monaco) => { editor.bar... }

   return (
      <MonacoEditor
         beforeMount={handleEditorWillMount}
         onMount={handleEditorDidMount}
      />
   )
}
```

Unfortunately, I personally still needed to use the `import type * as MonacoTypes` hack for the `IStandaloneThemeData` type. Sigh. Needs more types to be exported. /=
